app/controllers/application_controller.rb

```diff
@@ -45,4 +47,11 @@ def check_entitled(entitlement)
   def pundit_user
     UserContext.current_user_context
   end
+
+  def user_not_authorized(exception)
+    custom_message = "You are not authorized to #{exception.query.delete_suffix('?')} this #{exception.record.model_name.human.downcase}"
```

Comment:
@eclarizio Why not request this change in the Pundit Gem itself

Reply:
@mkanoor I looked, there was already a thread, they basically said to instead of returning false to raise the error itself and give it a reason. I didn't like that because I like that we can simply return true/false in the policy and have the error automatically generated. Maybe there's a better way for pundit to do it, but I also like that we can have control over the messages here if we need to do a `rescue_from` on other actions.

Or do you mean just the message in general? I *think* the `delete_suffix` method was added in Ruby 2.5, so that would be fine, but I'm almost positive that something like `record.model_name` and `.human` are both active record/active support, which I'm sure pundit doesn't want to depend on.

Reply:
@eclarizio 
The error message that we are trying to resolve happens during a create and for a Portfolio it looks like

not allowed to create?  this Class

it would have been ideal if we could have this read

not allowed to create?  this Portfolio

For a create operation we don't have an object so we end up passing the Portfolio class.

If we want to improve the error message we should just end up create a new Portfolio and pass that in so when it tries to log the message it appropriately.

https://github.com/varvet/pundit/blob/6be4621538f916124f76b57460c8cc46597a3b6a/lib/pundit.rb#L36

I don't think we are using the create pattern properly for Pundit. The authorize method in Pundit returns the record if Authorization was successful else it raises an error. So if we changed our create method in the controller to do this


```
      def create
        portfolio = authorize(Portfolio.new(writeable_params_for_create))
        portfolio.save!
        portfolio.reload
        render :json => portfolio
      end
```

instead of this
```
      def create
        authorize(Portfolio)

        portfolio = Portfolio.create!(writeable_params_for_create)
        render :json => portfolio
      end
``` 

This will change the exception message from

not allowed to create?  this Class
to
not allowed to create?  this Portfolio

as afar as the boolean methods with the ? that is a Rubyism which we are trying to expose to the outside world. In pundit the policy methods dont have to end with a ?. We add the ? to conform to the boolean method style and then remove them when we do user capabilities.

https://github.com/RedHatInsights/catalog-api/blob/0829c2bfafc4f5cd7d6cac4f7f34837663249505/app/policies/application_policy.rb#L36


If we just left them as create instead of create? that problem can be resolved.

Reply:
@mkanoor I agree, we could use `.new` and pass in the parameters and then pass that through to the authorize method. I don't think it's necessary though, the way I have it now handles both the create and the update case. I've updated the specs to show this.

Also, the boolean methods with the ? being a Rubyism is true, but pundit is explicitly checking for the definition of those with a question mark at the end https://github.com/varvet/pundit/blob/master/lib/pundit.rb#L218. Otherwise, we will have to provide the `query` optional parameter in every single authorize method:
`authorize(portfolio, :create)`.

That totally sucks that we would have to explicitly call the thing that pundit should be figuring out for us. It would be fine if we only had to define something like an admin check on everything, then we would save policy method writing but we'd again be duping `authorize(object, :admin_check)` everywhere.

Also I think you're completely ignoring the very first requirement of the JIRA story in that `Pundit::NotAuthorizedError` is appearing to the user, who has no idea what that means. Obviously, they can probably figure out why it says `NotAuthorizedError`, but as for why there is a `Pundit` showing up, they likely both don't care and have no use for it.

Reply:
@eclarizio 
The exception class name is something that can be removed in the common gem. None of those exception class names would make sense to the user.

https://github.com/RedHatInsights/insights-api-common-rails/blob/770bb10e8e681ddcc37c08545040a41aa0ed5a2a/lib/insights/api/common/application_controller_mixins/exception_handling.rb#L16

And this applies to all error messages that get displayed via the Error Document. Maybe the exception class name should stay in the log but not be transmitted to the user.

When I initially opened the issue in JIRA it was regarding the Class message which was completely misleading.

![Screenshot from 2020-03-10 14-32-18](https://user-images.githubusercontent.com/6452699/81228054-6424a180-8fbb-11ea-90a1-b4a5cc5f3f78.png)

If we need to fix the language of the exception I think we should open a PR against the pundit gem.

Also in this scenario the actual HTTP status is 403 which is going to be displayed in the Title in the error box based on the Javascript provided StatusText.
https://developer.mozilla.org/en-US/docs/Web/API/Response/statusText

So what we provide is additional details.

Reply:
I don't know if it's a good idea to try to get a change that only we really care about into an outside gem. I would imagine the vast majority of the people using pundit can simply use `rescue_from` and then give their own custom message. We're in a unique spot with common where common is rescuing everything for us leaving us with no control over the outcome.

I was able to prove locally that I *could* simply change everything in common, so I guess we can just do that.